Implement custom memory manager for LLVM

yuyichao opened the pull request:
Use various ways to reuse the page that has the page protection set in order to avoid wasting a page of memory for each JIT event.

Fixes #14626

Tested on Linux with LLVM 3.7.1 and 3.8 with OrcJIT.
This reduces the memory used in the subarray test by `~100MB` (There are around `25k` JIT events each page protects at least one page of memory even though the total allocation is only around `10-20MB`.)

Should in principle work on OSX, using the ~~`mkstemp`~~ `tmpfile` fallback code path.

~~Does not work on windows yet due to how we handle windows unwinding (we patch the code section after the code is generated, causing a segfault since the load address is never writable). It might be enough to modify those code to write to local address instead of load address. @Keno also mentioned some way to fix this properly. (This is what the request for help is mainly for....)~~ Fixed

yuyichao committed:
Move memory manager out of debuginfo.cpp

yuyichao committed:
Implement custom memory manager for LLVM

Use various ways to reuse the page that has the page protection set in order
to avoid wasting a page of memory for each JIT event.

Fixes #14626

yuyichao commented:
Another note is that LLVM 3.7.1 seems to generate corrupted eh_frame when using Keno's remote memory manager. Not sure why that happens but my current suspicion is the reuse of local address.

vtjnash commented:
> also mentioned some way to fix this properly. (This is what the request for help is mainly for....)

at present, this has been blocked on the need to write our own custom memory manager (specifically, one that supports the small memory model required by ntdll on win64 for all dynamic libraries), and then patch the LLVM runtime relocation code to add support for the ADDR32NB relocation type.

yuyichao commented:
> specifically, one that supports the small memory model required by ntdll on win64 for all dynamic libraries

What's the requirement? All sections to be smaller and within 4G to each other? The current custom memory manager doesn't do that yet but it's not that hard to add (especially if we can use the reserve function).

vtjnash commented:
yes, and consecutively allocated (images must be non-overlapping and can't be interleaved)

yuyichao commented:
> (images must be non-overlapping and can't be interleaved)

Doesn't this require the current allocation scheme? The page' where we have allocations in basically can't be reused anymore.

yuyichao commented:
Or maybe we can waste virtual memory but not physical memory?

vtjnash commented:
I'm not sure. It appears that the actual requirements are undocumented and considered to be an implementation detail. We may be able to get away with some lying, since we are JIT and not statically linked. But unfortunately, there's no way to verify the source code of ntdll to learn it's expectations.

Keno commented:
Why would there be an expectation that the sections are non-interleaved?

vtjnash commented:
Actually, maybe only dbghelp cares (https://msdn.microsoft.com/en-us/library/windows/desktop/ms681353(v=vs.85).aspx), so perhaps it would only cause problems for WinDbg, but not ntdll.

yuyichao commented:
:tada: All CI passed!!!

I had to change the windows allocation strategy since the windows memory management is even more stupid than I thought....

vtjnash commented:
Exciting! It looks like win64 retired 15 workers on this merge commit, instead of 17 :)

tkelman commented on src/cgmemmgr.cpp:
```diff
@@ -0,0 +1,839 @@
+//
+//   2. private dual map
+//
+//       Same with above but use anonymous memory map as memory pool,
```
Same as above

tkelman commented on src/cgmemmgr.cpp:
```diff
@@ -0,0 +1,839 @@
+// one page is wasted on each discarded blocks. There should be at most one
+// block with more than 128 pages available so the discarded one must have
+// less than 128 pages available and therefore at least 128 pages used.
+// (Apart from fragmentation) this guarantees less one 1% of memory wasting.
```
less than 1%

tkelman commented on src/cgmemmgr.cpp:
```diff
@@ -0,0 +1,839 @@
+    // * Initial allocation: `state & InitAlloc`
+    // * Followup allocation: `(state & Alloc) && !(state & InitAlloc)`
+    enum State {
+        // This block is has no page protection set yet.
```
~~is~~ has

tkelman commented on src/cgmemmgr.cpp:
```diff
@@ -0,0 +1,839 @@
+            return;
+        }
+        if (block.state & ROBlock::InitAlloc) {
+            // for a initial block, we need to map it to ro or rx
```
an initial

tkelman commented on src/cgmemmgr.cpp:
```diff
@@ -0,0 +1,839 @@
+        if (!(block.state & ROBlock::Alloc))
+            return;
+        if (block.state & ROBlock::InitAlloc) {
+            // for a initial block, we need to map it to ro or rx
```
an initial

tkelman commented on src/cgmemmgr.cpp:
```diff
@@ -0,0 +1,839 @@
+{
+    return new RTDyldMemoryManagerJL();
+}
+#endif
```
would be good to add comments on these w.r.t. what they're closing

Keno commented on src/cgmemmgr.cpp:
```diff
@@ -0,0 +1,839 @@
+// as a block on windows
+static void *create_shared_map(size_t size, size_t offset)
+{
+    void *addr = MapViewOfFile((HANDLE)offset, FILE_MAP_ALL_ACCESS,
```
Not sure how I feel about this punning.

yuyichao replied:
What do you mean?

Keno replied:
offset/handle

yuyichao replied:
Renamed to `id`.

Keno commented on src/cgmemmgr.cpp:
```diff
@@ -0,0 +1,839 @@
+    }
+};
+
+struct ROBlock : public Block {
```
`DualMappedBlock`?

yuyichao replied:
Sure, it's also used by the `/proc/self/mem` one for the flag though.

Keno replied:
`SplitPtrBlock` or something to that extent?

yuyichao replied:
Renamed to `SplitPtrBlock`

Keno commented on src/cgmemmgr.cpp:
```diff
@@ -0,0 +1,839 @@
+
+struct Allocation {
+    void *wr_addr;
+    void *rt_addr;
```
Why `rt` not `rd`?

yuyichao replied:
runtime?

yuyichao replied:
Keep as rt and comment added.

Keno commented on src/cgmemmgr.cpp:
```diff
@@ -0,0 +1,839 @@
+    {
+        if (exec) {
+            for (auto &alloc: allocations) {
+                sys::Memory::InvalidateInstructionCache(alloc.rt_addr,
```
@vtjnash says this isn't enough. Ask him for details.

yuyichao replied:
Not sure what's more is needed for this one still. @vtjnash

Keno commented on src/cgmemmgr.cpp:
```diff
@@ -0,0 +1,839 @@
+    ROBlock blocks[nblocks];
+    // Blocks that are done allocating but might not have all the permissions
+    // set or data copied yet.
+    SmallVector<ROBlock, 16> pendings;
```
This could be cleared up a bit. We didn't realize that these had been removed from blocks. Also, maybe there's a better name.

yuyichao replied:
`to_be_removed`?

Keno replied:
`completed`?

yuyichao replied:
Renamed.

Keno commented on src/cgmemmgr.cpp:
```diff
@@ -0,0 +1,839 @@
+    void finalize_block(ROBlock &block, bool reset)
+    {
+        if (!(block.state & ROBlock::Alloc)) {
+            if ((block.state & ROBlock::WRInit) && reset)
```
Does this ever happen?

yuyichao replied:
I'm not sure what llvm guarantees. This will happen if we are called with a large block that happens to have less free space then other blocks. (e.g. all other blocks has 1k free space and LLVM asked for a 1023.5k block followed by a 2k block).

yuyichao replied:
In another word, if LLVM guarantees that no more than one allocation will be made before each finalization, then this won't happen.

Keno replied:
What I guess I'm not sure about is why there wouldn't be an allocation, but the write pointer is initialized.

yuyichao replied:
Ah, you are right, this was left over of an old logic when I'm not setting Alloc for the initial one. I'll fix it.

yuyichao replied:
Actually this can happen (I mixed up the branch condition and the assert condition below) and the case I described is the third branch below. I'll just add more comment to this...

`!Alloc` means that this page is not used this time but it could be used multiple times before and therefore has the write address mapped. (The write address is only `mprotect(PROT_READ)` in finalization so it will still consume address space if we don't unmap it)

Keno commented on src/cgmemmgr.cpp:
```diff
@@ -0,0 +1,839 @@
+            // for an initial block, we need to map it to ro or rx
+            assert(!(block.state & (ROBlock::WRReady | ROBlock::WRInit)));
+            protect_page(block.ptr, block.total, exec ? Prot::RX : Prot::RO);
+            block.state = 0;
```
Why is this not set to `ROBlock::Alloc`?

yuyichao replied:
`ROBlock::Alloc` means the block is touched after last finalization (otherwise we don't need to set any mprotect on it.)

Keno replied:
Adjust the description of the flag accordingly?

Keno commented on src/cgmemmgr.cpp:
```diff
@@ -0,0 +1,839 @@
+    void operator=(const RTDyldMemoryManagerJL&) = delete;
+    SmallVector<EHFrame, 16> pending_eh;
+    RWAllocator rw_alloc;
+    ROAllocator<false> *ro_alloc;
```
std::unique_ptr

yuyichao replied:
=)

yuyichao replied:
Done

Keno commented on src/cgmemmgr.cpp:
```diff
@@ -0,0 +1,839 @@
+        pending_eh.clear();
+        return false;
+    }
+    else {
```
assert(!exe_allocator).

yuyichao replied:
Maybe just add this to the constructor?

Keno replied:
Maybe, but I like it here, since that's the implicit assumption.

Keno commented on src/cgmemmgr.cpp:
```diff
@@ -0,0 +1,839 @@
+
+#ifdef LLVM38
+void RTDyldMemoryManagerJL::notifyObjectLoaded(RuntimeDyld &Dyld,
+                                               const object::ObjectFile &Obj)
```
override

Keno replied:
Oh, sorry, this is out of line.

yuyichao replied:
To be fair, I think I actually had override here initially without realizing that it is only allowed in the declaration......

Keno commented on src/codegen_internal.h:
```diff
@@ -28,8 +28,12 @@ extern bool jl_dylib_DI_for_fptr(size_t pointer, const object::ObjectFile **obje
+                                         RTDyldMemoryManager *memmgr);
+#ifdef _OS_WINDOWS_
+void *lookupWriteAddressFor(RTDyldMemoryManager *memmgr, void *rt_addr);
+#endif
```
notifyObjectLoaded here as well from < 3.8

yuyichao replied:
OK.... I left it out so that I don't have to figure out which header to use for `llvm::orc::ObjectLinkingLayerBase::ObjSetHandleT` ...

Keno commented on src/jitlayers.cpp:
```diff
@@ -129,8 +131,17 @@ static void addOptimizationPasses(T *PM)
 }
 
+#ifdef USE_MCJIT
+RTDyldMemoryManager* createRTDyldMemoryManager(void);
```
Move this to codegen_internal.h as well.

yuyichao commented on src/debuginfo.cpp:
```diff
@@ -400,25 +402,37 @@ class JuliaJITEventListener: public JITEventListener
+#ifdef USE_ORCJIT
+                if (memmgr)
+                    SectionAddr =
+                        (uintptr_t)lookupWriteAddressFor(memmgr,
```
@Keno BTW, any idea why the `SectionAddr` and `SectionLoadAddr` are always the same without this?

yuyichao commented:
I think I addressed all the comments apart from the one about cache flushing. I was hoping that `mprotect` will flush the write to the memory and calling the LLVM helper will flush the instruction cache for the runtime address. Not sure what's still missing.

Keno commented:
@yuyichao I think this should be merged. Will you do the honor?

yuyichao commented:
Sure, I was just waiting for one of you to merge it ;-p

Keno commented:
This may be triggering a compiler bug in gcc 4.8, complaining that the copy constructor of Block was deleted (even though the use in question should have been a move constructor). We have have to comment out that deletion for the GCC 4.8 series.

yuyichao commented:
Where does it need a copy constructor for `Block` (whether move or not)? `temp_buff.emplace_back()`?

tkelman commented:
See buildbot log if it's more useful - https://build.julialang.org/builders/build_ubuntu14.04-x64/builds/940/steps/shell_1/logs/stdio
also quick to reproduce / test things since it only takes a couple minutes to hit.

Keno commented:
`SmallVector`.

yuyichao commented:
That one is supposed to use a placement new with the default constructor... I guess the LLVM 3.7 `SmallVector` implementation might be constructing a Block and use the placement new to move..... GCC 6.1.1 didn't complain though......

Keno commented:
What default constructor would it be using? How would it get the Block into its storage without using the move constructor?

yuyichao commented:
`new (ptr) Block()`? The default constructor with no argument is provided.

Keno commented:
So you were relying on that particular function not being instantiated because you're not using it?

yuyichao commented:
And I want to make sure I don't copy/move it accidentally since it can cause memory leak or memory free'd at the wrong time.

Keno commented:
Make the move constructor `assert(false)`?

yuyichao commented:
Actually moving construct is always fine, moving assignment is not and is deleted. I've updated https://github.com/JuliaLang/julia/pull/16950 to implement the moving constructor properly.

tkelman commented on src/cgmemmgr.cpp:
```diff
@@ -0,0 +1,860 @@
+
+static int init_self_mem()
+{
+    int fd = open("/proc/self/mem", O_RDWR | O_SYNC | O_CLOEXEC);
```
need to check before using `O_CLOEXEC` https://build.julialang.org/builders/package_tarball64/builds/475/steps/make%20binary-dist/logs/stdio

yuyichao replied:
That kernel is actually too old for this method to work although I guess we should still compile this method for this to be included in the generic binary.... Should be fixed in https://github.com/JuliaLang/julia/commit/3f4e1e7274993b32b4dab8bdabdc9b645993923c

JeffBezanson commented:
BTW, the memory savings from this is absolutely great. Awesome work @yuyichao !
